source/extensions/filters/network/postgres_proxy/postgres_decoder.cc

```diff
@@ -176,134 +179,216 @@ void DecoderImpl::initialize() {
   };
 }
 
-Decoder::Result DecoderImpl::parseHeader(Buffer::Instance& data) {
-  ENVOY_LOG(trace, "postgres_proxy: parsing message, len {}", data.length());
+/* Main handler for incoming messages. Messages are dispatched based on the
+   current decoder's state.
+*/
+Decoder::Result DecoderImpl::onData(Buffer::Instance& data, bool frontend) {
+  switch (state_) {
+  case State::InitState:
+    return onDataInit(data, frontend);
+  case State::OutOfSyncState:
+  case State::EncryptedState:
+    return onDataIgnore(data, frontend);
+  case State::InSyncState:
+    return onDataInSync(data, frontend);
+  default:
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  }
+}
+
+/* Handler for messages when decoder is in Init State. There are very few message types which
+   are allowed in this state.
+   If the initial message has the correct syntax and  indicates that session should be in
+   clear-text, the decoder will move to InSyncState. If the initial message has the correct syntax
+   and indicates that session should be encrypted, the decoder stays in InitState, because the
+   initial message will be received again after transport socket negotiates SSL. If the message
+   syntax is incorrect, the decoder will move to OutOfSyncState, in which messages are not parsed.
+*/
+Decoder::Result DecoderImpl::onDataInit(Buffer::Instance& data, bool) {
+  ASSERT(state_ == State::InitState);
 
   // The minimum size of the message sufficient for parsing is 5 bytes.
-  if (data.length() < 5) {
+  if (data.length() < 4) {
```

Comment:
What about create something like `#define MIN_MESSAGE_LENGTH 4` ? And seems the comment above is not in sync what code does...